drivers/SAI.cpp

```diff
@@ -0,0 +1,71 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2013 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "drivers/SAI.h"
+#include "platform/mbed_critical.h"
+
+#if DEVICE_SAI
+
+namespace mbed {
+
+SAI::SAI(const sai_format_t *fmt, bool is_input, uint32_t master_clock, bool internal_mclk) : _sai(), _mutex(), _is_input(is_input) {
+    // No lock needed in the constructor
+    sai_init_t init = {};
+    init.is_receiver = is_input;
+    init.is_slave = false;
+
+    init.mclk = SAI_MCLK;
+    if (_is_input) {
+        init.sd = SAI_RX_SD;
+        init.bclk = SAI_RX_BCLK;
+        init.wclk = SAI_RX_WCLK;
+    } else {
+        init.sd = SAI_TX_SD;
+        init.bclk = SAI_TX_BCLK;
+        init.wclk = SAI_TX_WCLK;
+    }
+
+    init.mclk_internal_src = internal_mclk;
+    init.mclk_freq = master_clock;  // 0 means guess it by yourself, else apply a clock div
+    init.sample_rate = 16000;
+    init.master_clock = 384 * init.sample_rate;
```

Comment:
Sample rate should be configurable.
The magic number used to define master clock (serial clock) is also unclear. I think that it should be defined based on sampling frequency, data format (number of transmitted bits) and number of channels (mono, stereo).

Reply:
100% Agreed
That's why I say that the C++ API is not to be considered final.

This PR is about the HAL API only (`hal/sai_api.h` and `hal/mbed_sai_api.c`).
But to be able to provide a working [demo](https://github.com/ithinuel/mbed-os-example-sai), this CPP layer was kind of mandatory.

I could have called the hal api directly from the demo but this would have been less _idiomatic_ from mbed's perspective.